### ISSUE TITLE
Error when subscribing to an API stream over TLS

### ISSUE THREAD

**Issue by avive** [spacemeshos/go-spacemesh#2235]
1. Try:
```
$ grpcurl api-20.spacemesh.io:443 spacemesh.v1.MeshService.LayerStream
```
Result:
```
ERROR:
  Code: Unavailable
  Message: Gateway Timeout: HTTP status code 504; transport: received the unexpected content-type "text/html"
```
2. Try again - it now works as expected and the data is streamed to console.

So I suspect it is a configuration issue and related to TLS as this is not happening when connecting insecurely to localhost.

@valar999

**Comment by valar999** [comment]
Looks like we have timeout in `grpcurl` itself. Connection closing has been initiated from client.
```
$ grpcurl api-20.spacemesh.io:443 spacemesh.v1.MeshService.LayerStream
ERROR:
  Code: Unavailable
  Message: Gateway Timeout: HTTP status code 504; transport: received the unexpected content-type "text/html"
```

```
01:29:07.688703 IP 35.187.98.56.443 > 192.168.1.16.64916: Flags [P.], seq 9961:10152, ack 1008, win 338, options [nop,nop,TS val 411138374 ecr 525759046], length 191
01:29:07.688843 IP 192.168.1.16.64916 > 35.187.98.56.443: Flags [.], ack 9961, win 2045, options [nop,nop,TS val 525819234 ecr 411138283], length 0
01:29:07.688846 IP 192.168.1.16.64916 > 35.187.98.56.443: Flags [.], ack 10152, win 2042, options [nop,nop,TS val 525819234 ecr 411138283], length 0
01:29:07.688847 IP 192.168.1.16.64916 > 35.187.98.56.443: Flags [.], ack 10152, win 2042, options [nop,nop,TS val 525819234 ecr 411138374,nop,nop,sack 1 {9961:10152}], length 0
01:29:07.689412 IP 192.168.1.16.64916 > 35.187.98.56.443: Flags [P.], seq 1008:1073, ack 10152, win 2048, options [nop,nop,TS val 525819234 ecr 411138374], length 65
01:29:07.705914 IP 192.168.1.16.64916 > 35.187.98.56.443: Flags [F.], seq 1073, ack 10152, win 2048, options [nop,nop,TS val 525819251 ecr 411138374], length 0
```

**Comment by avive** [comment]
We need to specify recommended client timeout and other settings for consuming grpc streams. This important for both smapp and cliwallet. lane is right and this issue should be moved to go-sm

**Comment by brusherru** [comment]
> Connection closing has been initiated from client.

This is not a truth at least at the moment.

I dig it a little bit using `grpcurl` and Wireshark to sniff the traffic.
Fortunately grpcurl tool supports SSLKEYLOGFILE so Wireshark can sniff encoded traffic. To do so you just need to set env variable `export SSLKEYLOGFILE=/some/path/keylog.txt` in the terminal session before calling `grpcurl` and then specify this file in Wireshark settings "Protocols > TLS".

So, how to reproduce and ensure that the client don't do anything wrong:
```
grpcurl -d '{ "transactionId": [{ "id": "MHhiN2ZjYTY4ZTM5OGQ5YjJmNTU5NGQyMDYyYTg2ZDJlNzQzNzRhNTkyNmU5MGIxZmVjZjQ3ZWY3NTY4Y2FkNWQ4" }] }' api-devnet.spacemesh.io:443 spacemesh.v1.TransactionService.TransactionsStateStream
```
(Looking ahead, I can tell that it works the same with `-keealive-time 5` and `-max-time 180` options)

And what I got in the Wireshark:
```
No	Time	From	To	Protocol	Length	Info
...
/// SEND REQUEST
154	2.354323	10.0.1.5	35.238.216.226	HTTP2	214	HEADERS[1]: POST /grpc.reflection.v1alpha.ServerReflection/ServerReflectionInfo
155	2.355592	10.0.1.5	35.238.216.226	GRPC	135	DATA[1] (GRPC) (PROTOBUF)
...
/// GOT AN EXPECTED RESPONSE
170	2.502501	35.238.216.226	10.0.1.5	HTTP2	1368	HEADERS[1]: 200 OK
171	2.502503	35.238.216.226	10.0.1.5	HTTP2	156	DATA[1] [HTTP2 segment of a reassembled PDU]
173	2.502507	35.238.216.226	10.0.1.5	GRPC	161	DATA[1] (GRPC) (PROTOBUF)
...
/// SERVER RESPONSES WITH 504
5477	62.653179	35.238.216.226	10.0.1.5	HTTP2	199	RST_STREAM[1], HEADERS[3]: 504 Gateway Time-out
5478	62.653184	35.238.216.226	10.0.1.5	HTTP2	257	DATA[3] (text/html)
...
/// AND CLIENT SENDS RST_STREAM 3
5480	62.653633	10.0.1.5	35.238.216.226	HTTP2	131	RST_STREAM[3], WINDOW_UPDATE[0], PING[0]
```
I left only related things, but you can [download entire logs](https://github.com/spacemeshos/go-spacemesh/files/7475684/txStream.pcapng.zip)

I've already found some issues in the internet related to nginx proxy timeout, but didn't found a solution without timeout at all 🤔 

Without nginx in the middle (running on the local node) Streams works well.

We need to handle it to make Smapp works well in Wallet only mode.

cc: @moshababo @narayanprusty @avive

**Comment by moshababo** [comment]
Pending @brusherru to test it again.

**Comment by brusherru** [comment]
@moshababo @narayanprusty I can confirm that right now it is working pretty well.
However, from time to time (but always I had to wait a really long) TIMEOUT error occurs on Streams. But I didn't find out the way to reproduce it or to determine why it happens (it has different intervals). So I propose to postpone this thing for the "better times" and do not stuck on it.
To avoid experiencing a stuck state of Smapp in Wallet Only mode I added a feature (which is required anyway) that will restart GRPC streams in case of some errors.
